### PR TITLE
api: Adding Zipkin Tracing support

### PR DESCRIPTION
**Pull request by alexandermarston**
**What type of PR is this?**
API to support Zipkin provider for Tracing telemetry.

**What this PR does / why we need it**:
At present, Envoy Gateway only supports the OpenTelemetry tracing provider.

This PR updates the underlying APIs to also support the Zipkin provider.

In addition, it also exposes a new option to specify provider specific configuration such as TraceId128Bit and SharedSpanContext for Zipkin.

**Which issue(s) this PR fixes**:
Related to #1827

https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/trace/v3/zipkin.proto

### CODE REVIEW

File: api/v1alpha1/tracing_types.go [hunk 1]
```diff
@@ -114,3 +116,18 @@ type RequestHeaderCustomTag struct {
 	// +optional
 	DefaultValue *string `json:"defaultValue,omitempty"`
 }
+
+// TracingProviderSettings defines optional configuration for tracing providers.
+type TracingProviderSettings struct {
```
Comment (zirain) [inline api/v1alpha1/tracing_types.go:121]: IIRC, these configurations are zipkin only, so it should under TracingProvider like following:
```
provider:
  type: Zipkin
  zipkin:
     traceId128Bit:
     sharedSpanContext:
```

**Comment by zirain** [comment]
please revert changes under `/internal`

File: api/v1alpha1/tracing_types.go (hunk 1 above)
Comment (alexandermarston) [reply api/v1alpha1/tracing_types.go]: Agree - I have moved things around.

**Comment by alexandermarston** [comment]
> please revert changes under `/internal`

@zirain I have done so - shall I update the site docs? (site/content/en/latest/api/extension_types.md)

File: api/v1alpha1/tracing_types.go [hunk 2]
```diff
@@ -58,6 +57,10 @@ type TracingProvider struct {
 	// +kubebuilder:validation:XValidation:message="only support Service kind.",rule="self.all(f, f.kind == 'Service')"
 	// +kubebuilder:validation:XValidation:message="BackendRefs only supports Core group.",rule="self.all(f, f.group == '')"
 	BackendRefs []BackendRef `json:"backendRefs,omitempty"`
+	// Zipkin defines optional configuration for the Zipkin tracing provider.
+	// +optional
+	// +kubebuilder:validation:XValidation:message="only supports Zipkin provider type.",rule="self.type == 'Zipkin'"
+	Zipkin ZipkinConfiguration `json:"zipkin,omitempty"`
```
Comment (zirain) [inline api/v1alpha1/tracing_types.go:63]:
```suggestion
	Zipkin *ZipkinConfiguration `json:"zipkin,omitempty"`
```

File: api/v1alpha1/tracing_types.go [hunk 3]
```diff
@@ -114,3 +117,16 @@ type RequestHeaderCustomTag struct {
...
+type ZipkinConfiguration struct {
+	// TraceId_128Bit determines whether a 128bit trace id will be used
+	// when creating a new trace instance.
+	// +optional
+	TraceId128Bit bool `json:"traceId128Bit,omitempty"`
+	// SharedSpanContext determines whether client and server spans will
+	// share the same span context. Defaults to true.
+	// +kubebuilder:default=true
+	// +optional
+	SharedSpanContext bool `json:"sharedSpanContext,omitempty"`
```
Comment (zirain) [inline api/v1alpha1/tracing_types.go:131]: should we rename this to DisableSharedSpanContext and make the default value to false?

**Comment by zirain** [comment]
> > please revert changes under `/internal`
> 
> @zirain I have done so - shall I update the site docs? (site/content/en/latest/api/extension_types.md)

run `make -k gen-check ` will auto generated the docs.

File: api/v1alpha1/tracing_types.go [hunk 4]
```diff
@@ -114,3 +117,16 @@ type RequestHeaderCustomTag struct {
 	// +optional
 	DefaultValue *string `json:"defaultValue,omitempty"`
 }
+
+// ZipkinConfiguration defines optional configuration for the Zipkin tracing provider.
+type ZipkinConfiguration struct {
+	// TraceId_128Bit determines whether a 128bit trace id will be used
```
[the remark this answers is not in the document]
Comment (alexandermarston) [reply api/v1alpha1/tracing_types.go:123]: _The default value is false, which will result in a 64 bit trace id being used._

https://pkg.go.dev/github.com/envoyproxy/go-control-plane/envoy/config/trace/v3@v0.12.1-0.20240612043845-c54ec4ce422d#ZipkinConfig.TraceId_128Bit

File: api/v1alpha1/tracing_types.go (hunk 3 above)
Comment (alexandermarston) [reply api/v1alpha1/tracing_types.go:131]: Seeing as SharedSpanContext defaults to true in the backend anyway, I think it makes sense to keep the naming of this field consistent.

Comment (zirain) [reply api/v1alpha1/tracing_types.go:130]: An API should imporve the UX of most of the end users, envoy may use specific values due to backward compatibility.

Comment (alexandermarston) [reply api/v1alpha1/tracing_types.go]: Updated

File: api/v1alpha1/tracing_types.go [hunk 5]
```diff
@@ -114,3 +116,20 @@ type RequestHeaderCustomTag struct {
...
+type ZipkinTracingProvider struct {
+	// CollectorHostname defines the hostname to use when sending spans
+	// to the Zipkin collector endpoint.
+	// +optional
+	CollectorHostname *string `json:"collectorHostname,omitempty"`
+	// TraceID128Bit determines whether a 128bit trace id will be used
+	// when creating a new trace instance. If set to false, a 64bit trace
+	// id will be used.
+	// +optional
+	TraceID128Bit bool `json:"traceId128Bit,omitempty"`
```
Comment (zirain) [inline api/v1alpha1/tracing_types.go:130]:
```suggestion
	TraceID128Bit *bool `json:"traceId128Bit,omitempty"`
```

let's follow the API guide, if it's optional, use a point.

File: api/v1alpha1/tracing_types.go [hunk 6]
```diff
@@ -114,3 +116,20 @@ type RequestHeaderCustomTag struct {
...
+	// DisableSharedSpanContext determines whether the default Envoy behaviour of
+	// client and server spans sharing the same span context should be disabled.
+	// +optional
+	DisableSharedSpanContext bool `json:"disableSharedSpanContext,omitempty"`
```
Comment (zirain) [inline api/v1alpha1/tracing_types.go:134]:
```suggestion
	DisableSharedSpanContext *bool `json:"disableSharedSpanContext,omitempty"`
```

File: api/v1alpha1/tracing_types.go (hunk 5 above)
Comment (alexandermarston) [reply api/v1alpha1/tracing_types.go]: Updated

File: api/v1alpha1/tracing_types.go (hunk 6 above)
Comment (alexandermarston) [reply api/v1alpha1/tracing_types.go]: Updated

File: api/v1alpha1/tracing_types.go [hunk 7]
```diff
@@ -114,3 +116,20 @@ type RequestHeaderCustomTag struct {
 	// +optional
 	DefaultValue *string `json:"defaultValue,omitempty"`
 }
+
+// ZipkinTracingProvider defines the Zipkin tracing provider configuration.
+type ZipkinTracingProvider struct {
+	// CollectorHostname defines the hostname to use when sending spans
+	// to the Zipkin collector endpoint.
+	// +optional
+	CollectorHostname *string `json:"collectorHostname,omitempty"`
```
Comment (zirain) [inline api/v1alpha1/tracing_types.go:125]: Is this necessary for first iterate? There's a discussion about where's the right place for authoriy, here or under `BackendRef`

Comment (alexandermarston) [reply api/v1alpha1/tracing_types.go:125]: It's not necessary but I lacked the context on the other discussion. Happy to pull this out and wait for the outcome of that discussion

File: api/v1alpha1/tracing_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (basvanbeek) [reply api/v1alpha1/tracing_types.go]: Most libraries do 64bit as default allowing to be switched to 128 bit for those worried about trace collisions in very high request rate systems, or where retention is set to high.

File: api/v1alpha1/tracing_types.go (hunk 7 above)
Comment (basvanbeek) [reply api/v1alpha1/tracing_types.go:125]: The standard Zipkin ecosystem does nothing with this.

**Review by zirain — DISMISSED** [review]
LGTM, @basvanbeek can you take a look?

**Review by basvanbeek — APPROVED** [review]
LGTM

File: api/v1alpha1/tracing_types.go [hunk 8]
```diff
@@ -114,3 +116,16 @@ type RequestHeaderCustomTag struct {
...
 	DefaultValue *string `json:"defaultValue,omitempty"`
 }
+
+// ZipkinTracingProvider defines the Zipkin tracing provider configuration.
+type ZipkinTracingProvider struct {
+	// TraceID128Bit determines whether a 128bit trace id will be used
+	// when creating a new trace instance. If set to false, a 64bit trace
+	// id will be used.
+	// +optional
+	TraceID128Bit *bool `json:"traceId128Bit,omitempty"`
```
Comment (arkodg) [inline api/v1alpha1/tracing_types.go:126]: I understand this is creating a 1:1 mapping with envoy https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/trace/v3/zipkin.proto#extension-envoy-tracers-zipkin but this name is confusing
prefer if this was `TraceIDSize/TraceIDLength *TraceIDType` which could default to `64`

Comment (alexandermarston) [reply api/v1alpha1/tracing_types.go:126]: Would `EnableTraceID128Bit` or `Enable128BitTraceID` be a better name?

Comment (arkodg) [reply api/v1alpha1/tracing_types.go:126]: yeah I'm a +1 for `Enable128BitTraceID` wdyt @envoyproxy/gateway-maintainers @basvanbeek

Comment (basvanbeek) [reply api/v1alpha1/tracing_types.go:126]: `Enable128BitTraceID` has my preference too.

Comment (alexandermarston) [reply api/v1alpha1/tracing_types.go]: Thank you for input. Field name updated

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Review by basvanbeek — APPROVED** [review]

**Review by zirain — APPROVED** [review]

**Pull request merged**